tests/test_aggregates.py

```diff
@@ -1,362 +1,375 @@
+        def keyfunction(job):
+            return job.sp["even"]
+
+        aggregate_store = self.create_aggregate_store(
+            aggregator.groupby(keyfunction), project
+        )
+        assert len(aggregate_store) == 2
+        for aggregate in aggregate_store.values():
+            even = aggregate[0].sp.even
+            assert all(even == job.sp.even for job in aggregate)
+
+    def test_groupby_with_invalid_callable_key(self, project):
+        def keyfunction(job):
+            return job.sp["half"]
+
+        with pytest.raises(KeyError):
+            # We will attempt to generate aggregates but will fail in
+            # doing so due to the invalid key
+            self.create_aggregate_store(aggregator.groupby(keyfunction), project)
+
+    def test_valid_select(self, project):
+        def _select(job):
+            return job.sp.i > 5
+
+        aggregate_store = self.create_aggregate_store(
+            aggregator.groupsof(1, select=_select), project
+        )
+        selected_jobs = []
+        for job in project:
+            if _select(job):
+                selected_jobs.append((job,))
+        assert self.get_aggregates_from_store(aggregate_store) == tuple(selected_jobs)
+
+    def test_store_hashing(self, project):
+        # Since we need to store groups on a per aggregate basis in the project,
+        # we need to be sure that the aggregates are hashing and compared correctly.
+        # This test ensures this feature.
+        total_aggregators = AggregateFixtures.list_of_aggregators()
+        list_of_stores = [
+            self.create_aggregate_store(aggregator, project)
+            for aggregator in total_aggregators
+        ]
+        assert len(list_of_stores) == len(total_aggregators)
+        # The above list contains 14 distinct store objects (because a
+        # store object is differentiated on the basis of the
+        # ``_is_default_aggregate`` attribute of the aggregator). When this
+        # list is converted to a set, then these objects are hashed first and
+        # then compared. Since sets don't carry duplicate values, we test
+        # whether the length of the set obtained from the list is equal to 14
+        # or not.
+        assert len(set(list_of_stores)) == 14
+
+    @pytest.mark.parametrize(
+        "aggregator_instance", AggregateFixtures.list_of_aggregators()
+    )
+    def test_aggregates_are_tuples(self, project, aggregator_instance):
+        # This test ensures that all aggregator functions return tuples. All
+        # aggregate stores are expected to return tuples for their values, but
+        # this also tests that the aggregator functions (groupsof, groupby) are
+        # generating tuples internally.
+        aggregate_store = self.create_aggregate_store(aggregator_instance, project)
+        if not isinstance(aggregate_store, _DefaultAggregateStore):
+            for aggregate in aggregate_store._generate_aggregates():

```

Comment:
On creation of the store, we register/compute the aggregates, so we don't need separate branches.